aws/aws-controllers-k8s
# E2E test framework written in Python

Issue #, if available: N/A

Description of changes:
The aim of this PR is to migrate our integration tests away from pure Bash scripts. Instead, I have written the existing SageMaker integration tests using PyTest. I have also provided a Dockerfile and a runner script (that builds and runs said Dockerfile) that encapsulates all of the dependencies and passes through the necessary variables to run on KIND (or EKS).

The flow of the integration test framework is as such:
- Run a bootstrap Python script (currently invoked using `./bootstrap <service>`)
  - Invokes the individual service's `service_bootstrap.py` file, creating AWS resources
  - Stores the exported properties of each resource into a `bootstrap.yaml` file in the service directory
- Run the PyTest module in the service directory.
  - Helper methods load templated resources (CR YAMLs) and injects values
  - Runs the PyTest fixtures and tests as expected
- Run a cleanup Python script (`./cleanup <service>`)
  - Invokes the individual service`s `service_cleanup`py` file, deleting the bootstrapped resources

Features:
- Tests can be marked as `canary`, if you wish to only run a subset of integration tests for canaries
- Helper methods provide abstractions over the K8s API for common tasks (those previously found in `lib/aws.sh`)

TODO:
- Add support for job parallelism (using `pytest-xdist`)
- Set up Github Actions workflow
- Support passing test markers through to Docker

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
